system/api-auth.go

```diff
@@ -0,0 +1,131 @@
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/decred/dcrd/dcrutil"
+	"github.com/decred/dcrwallet/wallet"
+	"github.com/dgrijalva/jwt-go"
+)
+
+// maxTicketChallengeAge is the maximum number of seconds into the past
+// or future required for a ticket auth timestamp value to be considered valid.
+const maxTicketChallengeAge = 60
+
+func (application *Application) validateToken(authHeader string) (int64, string) {
+	apitoken := strings.TrimPrefix(authHeader, "Bearer ")
+
+	JWTtoken, err := jwt.Parse(apitoken, func(token *jwt.Token) (interface{}, error) {
+		// validate signing algorithm
+		if _, ok := token.Method.(*jwt.SigningMethodHMAC); !ok {
+			return nil, fmt.Errorf("unexpected signing method: %v", token.Header["alg"])
+		}
+		return []byte(application.APISecret), nil
+	})
+
+	if err != nil {
+		return -1, fmt.Sprintf("invalid token %v: %v", apitoken, err)
+	} else if claims, ok := JWTtoken.Claims.(jwt.MapClaims); ok && JWTtoken.Valid {
+		return int64(claims["loggedInAs"].(float64)), ""
+	} else {
+		return -1, fmt.Sprintf("invalid token %v", apitoken)
+	}
+}
+
+func (application *Application) validateTicketOwnership(authHeader string) (multiSigAddress, authValidationFailureReason string) {
+	timestamp, timestampSignature, ticketHash := extractTicketAuthParams(strings.TrimPrefix(authHeader, "TicketAuth "))
+	if timestamp == "" || timestampSignature == "" || ticketHash == "" {
+		authValidationFailureReason = fmt.Sprintf("invalid ticket auth header value %s", authHeader)
+		return
+	}
+
+	// Ensure that this signature had not been used in a previous authentication attempt.
+	if application.ProcessedTicketChallenges.ContainsChallenge(timestampSignature) {
+		authValidationFailureReason = fmt.Sprintf("disallowed reuse of ticket auth signature %v", timestampSignature)
+		return
+	}
+
+	authTimestamp, err := strconv.Atoi(timestamp)
+	if err != nil {
+		authValidationFailureReason = fmt.Sprintf("invalid ticket auth timestamp value %v", timestamp)
+		return
+	}
+
+	// Ensure that the auth timestamp is not more than
+	// the permitted number of seconds into the past and future.
+	timestampDelta := time.Now().Unix() - int64(authTimestamp)
+	if math.Abs(float64(timestampDelta)) > maxTicketChallengeAge {
+		authValidationFailureReason = fmt.Sprintf("expired or invalid ticket auth timestamp value %v",
+			timestamp)
+		return
+	}
+
+	// confirm that the timestamp signature is a valid base64 string
+	decodedSignature, err := base64.StdEncoding.DecodeString(timestampSignature)
+	if err != nil {
+		authValidationFailureReason = fmt.Sprintf("invalid ticket auth signature %s", timestampSignature)
+		return
+	}
+
+	// Mark this timestamp signature as used to prevent subsequent reuse.
+	application.ProcessedTicketChallenges.AddChallenge(timestampSignature, maxTicketChallengeAge)
+
+	// todo check if ticket belongs to this vsp
+
+	// get user wallet address using ticket hash
+	// todo: may be better to maintain a memory map of tickets-userWalletAddresses
```

Comment:
Well, I wasn't actually thinking db per se, more like what we have here:
https://github.com/decred/dcrstakepool/blob/625bcd23ec3901d7b732903b9255a1a03104e0d2/backend/stakepoold/rpc/rpcserver/context.go#L39

Looking at that further, I see the data is initially read into memory from db directly here:
https://github.com/decred/dcrstakepool/blob/625bcd23ec3901d7b732903b9255a1a03104e0d2/backend/stakepoold/server.go#L196

However, since this information is so far only needed on dcrstakepool, saving it to db is a good idea.

If in the future, stakepoolds require this information, we can similar to how it's done above, query the db directly from stakepoold or as recommended in #495 (particularly https://github.com/decred/dcrstakepool/issues/495#issuecomment-523422290), we supply all db data required by stakepoold via RPC.

Reply:
I hope we could do that in a separate pr, this one is already pretty big.

Reply:
Good idea too. That PR could probably also resolve #495.